packages/ui-prop-types/src/controllable.ts

```diff
@@ -44,18 +55,35 @@
  * This will throw an error if the 'selected' prop is supplied without a corresponding
  * 'onSelect' handler and will recommend using 'defaultSelected' instead.
  * @module controllable
- * @param {function} propType - validates the prop type. Returns null if valid, error otherwise
+ * @param {function} checkType - validates the prop type. Returns null if valid, error otherwise
  * @param {string} handlerName - name of the handler function
  * @param {string} defaultPropName - name of the default prop
- * @returns {Error} if designated prop is supplied without a corresponding handler function
+ * @returns {function} A function that return Error if designated prop is
+ * supplied without a corresponding handler function
  */
-function controllable(
-  propType,
+function controllable<T>(
+  checkType: CheckTypeInternal & Requireable<T>,
   handlerName = 'onChange',
   defaultPropName = 'defaultValue'
 ) {
-  return function (props, propName, componentName) {
-    const error = propType.apply(null, arguments)
+  return function (
+    props: Record<string, any>,
+    propName: string,
+    componentName: string
+  ) {
+    // TODO this is a hack that should be removed.
+    // This calls an internal function that has a similar signature to
+    // prop-types/Requireable, but has an extra safeguard parameter, so its
+    // not called directly.
+    // See http://fb.me/use-check-prop-types
+    const error = checkType(
+      props,
+      propName,
+      componentName,
+      'prop',
+      propName,
+      'SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED'
+    )
```

Comment:
This was a nightmare to type. At first glance it looked like that `checkType` is the usual `Validator` type, but after some debugging and really weird bugs it turns out that the function passed to `checkType` is an internal function, that should not be called. React tries to prevent calling it by passing a secret last parameter to it, it was circumvented by using `arguments`. I just expanded this `arguments` so we can clearly see what gets passed.